### PR TITLE
Refinement: Removed Redundancy of Scan Config Passing

### PR DESCRIPTION
**Pull request by aryanagrawal22**
## Description

- Issue: https://github.com/google/gcp_scanner/issues/241. Hence I created a property `crawler_config` in base class `ICrawler` which is returns bool value, where `default=False`
- If there is a need or optionality of passing the config to the crawler then this property `crawler_config` in the base class needs to be overridden in the sub class to return True.

**Note** : Currently as I was checking all the crawlers, I have just seen that config is being used only inside the `StorageBucketsCrawler` crawler. So only its value is set as True in its base class.

## Changes Made

- Base class `ICrawler` has a new property called as `crawler_config` which returns the abstracted private variable `_config_depndency` which is set to **False** by default.
- `StorageBucketsCrawler` class has overridden  `crawler_config` to return True.
- `get_crawl` function in scanner.py has a check now calling the `crawler_config` to decide to send config or not.

## Testing

Local testing done with logs denoting that where config file is sent and where it is not. (**Removed the logs after the testing)**

Screenshot 2023-11-23 at 6 01 29 PM

This check denotes that only `StorageBucketsCrawler` is being sent the config file

### CODE REVIEW

File: src/gcp_scanner/crawler/interface_crawler.py [hunk 1]
```diff
@@ -26,6 +26,12 @@ class ICrawler(metaclass=ABCMeta):
 
   """
 
+  """Variable to identify if config file is needed
+
+  Access Type: Private
+  """
+  _config_depndency = False
```
Comment (mshudrak) [inline src/gcp_scanner/crawler/interface_crawler.py:33]: NIT: depndency/dependency. Here and everywhere else.

File: src/gcp_scanner/crawler/storage_buckets_crawler.py [hunk 2]
```diff
@@ -24,6 +24,8 @@
 class StorageBucketsCrawler(ICrawler):
   """Handle crawling of bucket names data."""
 
+  _config_depndency = True # Private Variable Denoting Config Files Needed
```
Comment (mshudrak) [inline src/gcp_scanner/crawler/storage_buckets_crawler.py:27]: I'd rephrase it to just "# Define that config file is needed"

File: src/gcp_scanner/scanner.py [hunk 3]
```diff
@@ -180,8 +180,10 @@ def get_crawl(
   Returns:
     scan_result: a dictionary with scanning results
   """
-
-  res = crawler.crawl(project_id, client, crawler_config)
+  if crawler.crawler_config:
```
Comment (mshudrak) [inline src/gcp_scanner/scanner.py:183]: You are not actually calling the function here. This statement will always return true.

**Review by mshudrak — CHANGES_REQUESTED** [review]
Thanks for your PR. I left some suggestions. PTAL.

File: src/gcp_scanner/crawler/interface_crawler.py (hunk 1 above)
Comment (aryanagrawal22) [reply src/gcp_scanner/crawler/interface_crawler.py]: Apologies, for the the Typo. Have fixed all its occurrences.

File: src/gcp_scanner/crawler/storage_buckets_crawler.py (hunk 2 above)
Comment (aryanagrawal22) [reply src/gcp_scanner/crawler/storage_buckets_crawler.py]: Updated the comment with the new commit.

File: src/gcp_scanner/scanner.py (hunk 3 above)
Comment (aryanagrawal22) [reply src/gcp_scanner/scanner.py:183]: If you are referencing about the crawler_config not having parentheses denoting a function, it is because the crawler_config method is a property, and properties in Python are typically accessed without parentheses.

Screenshot 2023-11-23 at 6 01 29 PM

The above testing image shows that, when the scanner is run, it is not always giving true for crawler_config. False for every class except of `StorageBucketsCrawler` (As for now only that class needs the config file).


Please let me know, if I have misunderstood your comment or something else was wrong.

**Comment by aryanagrawal22** [comment]
> Thanks for your PR. I left some suggestions. PTAL.

Thank you for going through my PR, and reviewing it. I think, I have done the relevant changes as asked. Please let me know in case any others are needed.

File: src/gcp_scanner/scanner.py (hunk 3 above)
Comment (mshudrak) [reply src/gcp_scanner/scanner.py:183]: Oh my bad, right. NVM. I'd rename the function to make it more clear of what is property about. Right now we have too many `crawler_config`s in the code.

**Review by mshudrak — CHANGES_REQUESTED** [review]
One small suggestion and we are good to go.

File: src/gcp_scanner/scanner.py (hunk 3 above)
Comment (aryanagrawal22) [reply src/gcp_scanner/scanner.py]: Yes, fair point. I have made the changes, changed the name to `has_config_dependency` from `crawler_config`

Comment (aryanagrawal22) [reply src/gcp_scanner/scanner.py]: > Yes, fair point. I have made the changes, changed the name to `has_config_dependency` from `crawler_config`

Pushed with the changes.

**Comment by aryanagrawal22** [comment]
> Yes, fair point. I have made the changes, changed the name to has_config_dependency from crawler_config

Pushed with the changes 😁
